File: src/main/scala/org/xbery/artbeams/common/mailer/config/MailerConfig.scala
package org.xbery.artbeams.common.mailer.config

import javax.inject.Inject
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Component
import org.xbery.artbeams.config.repository.ConfigRepository

/**
  * Configuration of Mailer.
  * @author Radek Beran
  */
@Component
class MailerConfig @Inject() (
  configRepository: ConfigRepository,
  @Value("${mailer.api-key}")
  val apiKey: String,
  @Value("${mailer.domain}")
  val domain: String,
) {
  lazy val from = configRepository.requireConfig("mailer.from")
}

File: src/main/kotlin/org/xbery/artbeams/common/mailer/config/MailerConfig.kt
package org.xbery.artbeams.common.mailer.config

import org.springframework.core.env.Environment
import org.springframework.stereotype.Component
import org.xbery.artbeams.config.repository.AppConfig

/**
 * Configuration of Mailer.
 * @author Radek Beran
 */
@Component
open class MailerConfig (
    private val appConfig: AppConfig,
    private val env: Environment
) {
    // apiKey and domain is directly populated from Heroku environment variables

    fun getApiKey(): String = requireNotNull(env.getProperty("MAILGUN_API_KEY")) {
        "MAILGUN_API_KEY environment variable is not set"
    }

    fun getDomain(): String = requireNotNull(env.getProperty("MAILGUN_DOMAIN")) {
        "MAILGUN_DOMAIN environment variable is not set"
    }

    fun getFrom(): String = appConfig.requireConfig("mailer.from")
}
